File: vripper-server/src/main/java/tn/mnlr/vripper/host/AcidimgHost.java
package tn.mnlr.vripper.host;

import java.util.ArrayList;
import java.util.List;
import lombok.extern.slf4j.Slf4j;
import org.apache.http.NameValuePair;
import org.apache.http.client.HttpClient;
import org.apache.http.client.entity.UrlEncodedFormEntity;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.client.protocol.HttpClientContext;
import org.apache.http.message.BasicNameValuePair;
import org.apache.http.util.EntityUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.w3c.dom.Document;
import org.w3c.dom.Node;
import tn.mnlr.vripper.exception.HostException;
import tn.mnlr.vripper.exception.XpathException;
import tn.mnlr.vripper.jpa.domain.Image;
import tn.mnlr.vripper.services.ConnectionService;
import tn.mnlr.vripper.services.HostService;
import tn.mnlr.vripper.services.HtmlProcessorService;
import tn.mnlr.vripper.services.XpathService;

@Service
@Slf4j
public class AcidimgHost extends Host {

  private static final String host = "acidimg.cc";
  private static final String CONTINUE_BUTTON_XPATH = "//input[@id='continuebutton']";
  private static final String IMG_XPATH = "//img[@class='centred']";

  private final ConnectionService cm;
  private final HostService hostService;
  private final XpathService xpathService;
  private final HtmlProcessorService htmlProcessorService;

  @Autowired
  public AcidimgHost(
      ConnectionService cm,
      HostService hostService,
      XpathService xpathService,
      HtmlProcessorService htmlProcessorService) {
    this.cm = cm;
    this.hostService = hostService;
    this.xpathService = xpathService;
    this.htmlProcessorService = htmlProcessorService;
  }

  @Override
  public String getHost() {
    return host;
  }

  @Override
  public String getLookup() {
    return host;
  }

  @Override
  public HostService.NameUrl getNameAndUrl(final Image image, final HttpClientContext context)
      throws HostException {

    Document doc = hostService.getResponse(image.getUrl(), context).getDocument();

    Node contDiv;
    try {
      log.debug(
          String.format(
              "Looking for xpath expression %s in %s", CONTINUE_BUTTON_XPATH, image.getUrl()));
      contDiv = xpathService.getAsNode(doc, CONTINUE_BUTTON_XPATH);
    } catch (XpathException e) {
      throw new HostException(e);
    }

    if (contDiv != null) {
      log.debug(String.format("Click button found for %s", image.getUrl()));
      HttpClient client = cm.getClient().build();
      HttpPost httpPost = cm.buildHttpPost(image.getUrl(), context);
      httpPost.addHeader("Referer", image.getUrl());
      List<NameValuePair> params = new ArrayList<>();
      params.add(new BasicNameValuePair("imgContinue", "Continue to your image"));
      try {
        httpPost.setEntity(new UrlEncodedFormEntity(params));
      } catch (Exception e) {
        throw new HostException(e);
      }

      log.debug(String.format("Requesting %s", httpPost));
      try (CloseableHttpResponse response =
          (CloseableHttpResponse) client.execute(httpPost, context)) {
        log.debug(String.format("Cleaning response for %s", httpPost));
        doc = htmlProcessorService.clean(EntityUtils.toString(response.getEntity()));
        EntityUtils.consumeQuietly(response.getEntity());
      } catch (Exception e) {
        throw new HostException(e);
      }
    }

    Node imgNode;
    try {
      log.debug(String.format("Looking for xpath expression %s in %s", IMG_XPATH, image.getUrl()));
      imgNode = xpathService.getAsNode(doc, IMG_XPATH);
    } catch (XpathException e) {
      throw new HostException(e);
    }

    if (imgNode == null) {
      throw new HostException("Cannot find the image node");
    }

    try {
      log.debug(String.format("Resolving name and image url for %s", image.getUrl()));
      String imgTitle = imgNode.getAttributes().getNamedItem("alt").getTextContent().trim();
      String imgUrl = imgNode.getAttributes().getNamedItem("src").getTextContent().trim();

      return new HostService.NameUrl(
          imgTitle.isEmpty() ? hostService.getDefaultImageName(imgUrl) : imgTitle, imgUrl);
    } catch (Exception e) {
      throw new HostException("Unexpected error occurred", e);
    }
  }
}

File: vripper-core/src/main/kotlin/me/vripper/host/AcidimgHost.kt
package me.vripper.host

import me.vripper.entities.ImageEntity
import me.vripper.exception.HostException
import me.vripper.exception.XpathException
import me.vripper.services.DownloadService.ImageDownloadContext
import me.vripper.services.HTTPService
import me.vripper.utilities.HtmlUtils
import me.vripper.utilities.LoggerDelegate
import me.vripper.utilities.XpathUtils
import org.apache.hc.client5.http.classic.methods.HttpPost
import org.apache.hc.client5.http.entity.UrlEncodedFormEntity
import org.apache.hc.core5.http.message.BasicNameValuePair
import org.w3c.dom.Node

internal class AcidimgHost(
    private val httpService: HTTPService,
) : Host("acidimg.cc", 0) {
    private val log by LoggerDelegate()

    @Throws(HostException::class)
    override fun resolve(
        image: ImageEntity, context: ImageDownloadContext
    ): Pair<String, String> {
        val document = fetchDocument(image.url, context)
        try {
            log.debug(
                String.format(
                    "Looking for xpath expression %s in %s", CONTINUE_BUTTON_XPATH, image.url
                )
            )
            XpathUtils.getAsNode(document, CONTINUE_BUTTON_XPATH)
        } catch (e: XpathException) {
            throw HostException(e)
        }
        log.debug(String.format("Click button found for %s", image.url))
        val httpPost = HttpPost(image.url).also {
            it.addHeader("Referer", image.url)
            it.entity = UrlEncodedFormEntity(
                listOf(
                    BasicNameValuePair(
                        "imgContinue",
                        "Continue to your image"
                    )
                )
            )
        }.also { context.requests.add(it) }
        log.debug(String.format("Requesting %s", httpPost.uri))
        val doc = try {
            httpService.client.execute(
                httpPost, context.httpContext
            )
            { response ->
                log.debug(String.format("Cleaning response for %s", httpPost))
                HtmlUtils.clean(response.entity.content)
            }
        } catch (e: Exception) {
            throw HostException(e)
        }
        val imgNode: Node = try {
            log.debug(String.format("Looking for xpath expression %s in %s", IMG_XPATH, image.url))
            XpathUtils.getAsNode(doc, IMG_XPATH)
        } catch (e: XpathException) {
            throw HostException(e)
        } ?: throw HostException(
            String.format(
                "Xpath '%s' cannot be found in '%s'", IMG_XPATH, image.url
            )
        )
        return try {
            log.debug(String.format("Resolving name and image url for %s", image.url))
            val imgTitle = imgNode.attributes.getNamedItem("alt").textContent.trim()
            val imgUrl = imgNode.attributes.getNamedItem("src").textContent.trim()
            Pair(
                imgTitle.ifEmpty { getDefaultImageName(imgUrl) }, imgUrl
            )
        } catch (e: Exception) {
            throw HostException("Unexpected error occurred", e)
        }
    }

    companion object {
        private const val CONTINUE_BUTTON_XPATH = "//input[@id='continuebutton']"
        private const val IMG_XPATH = "//img[@class='centred']"
    }
}